packages/plugins/response-cache/src/index.ts

```diff
@@ -22,35 +30,105 @@ function sessionFactoryForEnvelop({ request }: YogaInitialContext) {
   return operationIdByRequest.get(request)
 }
 
+const cacheKeyFactoryForEnvelop: BuildResponseCacheKeyFunction =
+  async function cacheKeyFactoryForEnvelop({ sessionId }) {
+    if (sessionId == null) {
+      throw new Error(
+        '[useResponseCache] This plugin is not configured correctly. Make sure you use this plugin with GraphQL Yoga',
+      )
+    }
+    return sessionId
+  }
+
+const getDocumentStringForEnvelop: GetDocumentStringFunction = (
+  executionArgs,
+) => {
+  const context = executionArgs.contextValue as YogaInitialContext
+  if (context.params?.query == null) {
+    throw new Error(
+      '[useResponseCache] This plugin is not configured correctly. Make sure you use this plugin with GraphQL Yoga',
+    )
+  }
+  return context.params.query as string
+}
+
 export function useResponseCache(options: UseResponseCacheParameter): Plugin {
   const buildResponseCacheKey: BuildResponseCacheKeyFunction =
     options?.buildResponseCacheKey || defaultBuildResponseCacheKey
   const cache = options.cache ?? createInMemoryCache()
   const enabled = options.enabled ?? (() => true)
+  const cachedLastModifiedByRequest = new WeakMap<Request, string>()
+  let logger: YogaLogger
   return {
+    onYogaInit({ yoga }) {
+      logger = yoga.logger
+    },
     onPluginInit({ addPlugin }) {
       addPlugin(
         useEnvelopResponseCache({
           ...options,
           cache,
+          getDocumentString: getDocumentStringForEnvelop,
           session: sessionFactoryForEnvelop,
+          buildResponseCacheKey: cacheKeyFactoryForEnvelop,
+          shouldCacheResult({ result }) {
+            const shouldCached = options.shouldCacheResult
+              ? options.shouldCacheResult({ result })
+              : !result.errors?.length
+            if (shouldCached) {
+              result.extensions ||= {}
+              result.extensions.responseCacheLastModified =
+                new Date().toString()
+            } else {
+              logger.warn('[useResponseCache] Failed to cache due to errors')
+            }
+            return shouldCached
+          },
         }),
       )
     },
+    async onRequest({ request, fetchAPI, endResponse }) {
+      if (enabled(request)) {
+        const operationId = request.headers.get('If-None-Match')
+        if (operationId) {
+          const cachedResponse = await cache.get(operationId)
+          if (cachedResponse) {
+            const lastModifiedFromClient =
+              request.headers.get('If-Modified-Since')
+            const lastModifiedFromCache = cachedResponse.extensions
+              ?.responseCacheLastModified as string
+            if (
+              new Date(lastModifiedFromClient!).getTime() >=
+              new Date(lastModifiedFromCache).getTime()
+            ) {
```

Comment:
Instead of the casts and ! we could simply do a type guard `typeof lastModifiedFromClient === "string" && typeof lastModifiedFromCache === "string"`?

Reply:
The header is always string actually.

Reply:
https://github.com/dotansimha/graphql-yoga/pull/2448

Reply:
`typeof` is needed only for TypeScript actually because `responseCacheTimestamp` is always there if I don't miss something 🤔 That's why I didn't do that.
Actually if `If-Modified-Since` doesn't exist, server should ignore the timestamp.

Reply:
then let's remove TypeScript from the project if it is wrong... 😁 I have rather a bit of runtime overhead than arbitrary type casts...

Reply:
Ok let's do it... Besides sarcasm, I mean it is a false positive that can be solved in a better way. `cache.get` signature can be fixed by adding these extensions. I'll try something.